File: backend/src/application/places/places.service.ts
import { Injectable, NotFoundException } from '@nestjs/common'
import { PrismaService } from '../../infra/prisma/prisma.service'
import type { CategoryId, PlaceCardDto, PlaceDetailDto } from '../../contracts/dto/place.dto'
import { PLACE_INCLUDE, toCard, toDetail } from './place.mapper'

@Injectable()
export class PlacesService {
  constructor(private readonly prisma: PrismaService) {}

  async list(cat?: CategoryId): Promise<PlaceCardDto[]> {
    const rows = await this.prisma.place.findMany({
      // public map shows ACTIVE places only — moderation INACTIVE/DRAFT hidden;
      // approving a submission publishes a new ACTIVE place that appears here.
      where: {
        status: 'ACTIVE',
        ...(cat ? { categories: { some: { categoryId: cat } } } : {}),
      },
      include: PLACE_INCLUDE,
      orderBy: { sort: 'asc' },
    })
    return rows.map(toCard)
  }

  async getBySlug(slug: string): Promise<PlaceDetailDto> {
    const row = await this.prisma.place.findUnique({
      where: { slug },
      include: PLACE_INCLUDE,
    })
    if (!row) throw new NotFoundException(`place not found: ${slug}`)
    return toDetail(row)
  }
}

File: api/src/main/java/ee/gemspot/api/service/PlacesService.java
package ee.gemspot.api.service;

import ee.gemspot.api.domain.Place;
import ee.gemspot.api.domain.PlaceStatus;
import ee.gemspot.api.dto.PlaceCardDto;
import ee.gemspot.api.dto.PlaceDetailDto;
import ee.gemspot.api.mapper.PlaceMapper;
import ee.gemspot.api.repository.PlaceRepository;
import org.springframework.http.HttpStatus;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;
import org.springframework.web.server.ResponseStatusException;

import java.util.List;

/**
 * Mirrors Nest PlacesService (backend/src/application/places/places.service.ts).
 * Public map shows ACTIVE places only; INACTIVE/DRAFT hidden.
 */
@Service
public class PlacesService {

    private final PlaceRepository placeRepo;
    private final PlaceMapper mapper;

    public PlacesService(PlaceRepository placeRepo, PlaceMapper mapper) {
        this.placeRepo = placeRepo;
        this.mapper = mapper;
    }

    @Transactional(readOnly = true)
    public List<PlaceCardDto> list(String cat) {
        List<Place> active = placeRepo.findByStatusOrderBySortAsc(PlaceStatus.ACTIVE);
        return active.stream()
                // cat filter mirrors Nest `categories: { some: { categoryId: cat } }`
                .filter(p -> cat == null || p.getCategories().stream()
                        .anyMatch(pc -> pc.getCategory().getId().equals(cat)))
                .map(mapper::toCard)
                .toList();
    }

    @Transactional(readOnly = true)
    public PlaceDetailDto getBySlug(String slug) {
        Place place = placeRepo.findBySlug(slug)
                .orElseThrow(() -> new ResponseStatusException(
                        HttpStatus.NOT_FOUND, "place not found: " + slug));
        return mapper.toDetail(place);
    }
}
